File: main_project/tripod2/java/core/src/main/java/xce/tripod/InvokeClient.java
package xce.tripod;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReentrantLock;



public class InvokeClient {

    private InvokeClient() {
        new Thread(task).start();
    }

    private static InvokeClient instance = null;

    static InvokeClient getInstance() {
        if (instance == null) {
            synchronized (InvokeClient.class) {
                if (instance == null) {
                    instance = new InvokeClient();
                }
            }
        }
        return instance;
    }

    /** 异步log任务 */
    private StatisticsLogTask task = new StatisticsLogTask();

    private ConcurrentHashMap<String, OperationInvokeStatistics> map = new ConcurrentHashMap<String, OperationInvokeStatistics>();

    private ReentrantLock lock = new ReentrantLock();

    private OperationInvokeStatistics getOpInvokeStatistics(String op, String namespaceId,
            String businessId) {
        String key = namespaceId + businessId + op;
        if (map.get(key) == null) {
            try {
                lock.lock();
                if (map.get(key) == null) {
                    map.put(key, new OperationInvokeStatistics(namespaceId, businessId, op, task));
                }
            } finally {
                lock.unlock();
            }
        }
        return map.get(key);
    }

    void log(String op, String namespaceId, String businessId, long count) {
        getOpInvokeStatistics(op, namespaceId, businessId).incCount(count);
    }

    void log(String op, String namespaceId, String businessId) {
        log(op, namespaceId, businessId, 1L);
    }
}

File: main_project/tripod2/cc/core/invoke_client.cc
#include "tripod2/cc/core/common.h"
#include "tripod2/cc/core/invoke_client.h"
#include "LogWrapper.h"

namespace xce {
namespace tripod {
 
InvokeClient::InvokeClient(): thread_(boost::bind(&StatisticsLogTask::Run, &task_)) {
}

 
InvokeClient::~InvokeClient() {
  MCE_INFO("InvokeClient::~InvokeClient()");
  Destroy();
  thread_.join();
}

void InvokeClient::Log(const std::string& op,const std::string& namespace_id,const std::string& biz_id, long count) {
  GetOperationInvokeStatistics(op,namespace_id,biz_id)->IncCount(count);
}

OperationInvokeStatisticsPtr InvokeClient::GetOperationInvokeStatistics(const std::string& op,const std::string& namespace_id,const std::string& biz_id) {
  std::string key = namespace_id + biz_id + op;
  MAP_TYPE::iterator iter; 
  if ( (iter=map_.find(key))  == map_.end() ) {
    boost::unique_lock<boost::mutex> lock(mutex_);
    if ( (iter=map_.find(key)) == map_.end() ) {
      map_.insert(make_pair(key,OperationInvokeStatisticsPtr(new OperationInvokeStatistics(namespace_id,biz_id,op,task_))));
    }
  }
  return map_[key];
}

InvokeClient* InvokeClient::instance_;
boost::mutex InvokeClient::singleton_mutex_;

void InvokeClient::Destroy() {
  MCE_INFO("InvokeClient::Destroy()");
  task_.Stop();
}

InvokeClient& InvokeClient::instance() {
  if (instance_ == NULL) {
    boost::unique_lock<boost::mutex> lock(singleton_mutex_);
    if (instance_ == NULL) {
      instance_ = new InvokeClient();
      // when helper destroys, instance_.Destroy will be invoked.
      static InvokeClientHelper helper(*instance_);
    }
  }
  return *instance_;
}

InvokeClientHelper::InvokeClientHelper(InvokeClient& handle) : handle_(handle) {
}

InvokeClientHelper::~InvokeClientHelper() {
  handle_.Destroy();
}

}} // xce::tripod



